File: langs/python3/problems/in_memory_database/stub.py
class InMemoryDatabase:
    def __init__(self):
        pass

    def set(self, key, field, value):
        """Write a field. Returns ""."""
        raise NotImplementedError

    def get(self, key, field):
        """Value, or "" if missing."""
        raise NotImplementedError

    def delete(self, key, field):
        """Returns "true" if deleted, "false" if missing."""
        raise NotImplementedError

    def scan(self, key):
        """Fields as field(value), ... sorted by field. Empty if none.

        Example: "abc(123), age(30), city(NY), name(Alice)".
        """
        raise NotImplementedError

    def scan_by_prefix(self, key, prefix):
        """Same as scan, only fields starting with prefix.

        Example: "abc(123), age(30)".
        """
        raise NotImplementedError

    def set_at(self, key, field, value, timestamp):
        """Write with no expiry. Returns ""."""
        raise NotImplementedError

    def set_at_with_ttl(self, key, field, value, timestamp, ttl):
        """Write live in [timestamp, timestamp + ttl). Returns ""."""
        raise NotImplementedError

    def delete_at(self, key, field, timestamp):
        """Delete if live at timestamp. "true" or "false"."""
        raise NotImplementedError

    def get_at(self, key, field, timestamp):
        """Value if live at timestamp, else ""."""
        raise NotImplementedError

    def scan_at(self, key, timestamp):
        """Live fields at timestamp as field(value), ..., or ""."""
        raise NotImplementedError

    def scan_by_prefix_at(self, key, prefix, timestamp):
        """Live fields with that prefix at timestamp, or ""."""
        raise NotImplementedError

    def backup(self, timestamp):
        """Snapshot live keys. Key count as a string."""
        raise NotImplementedError

    def restore(self, timestamp, timestamp_to_restore):
        """Load latest backup at or before timestamp_to_restore. Returns ""."""
        raise NotImplementedError

File: langs/java/problems/in_memory_database/stub.java
public class InMemoryDatabase {
    public InMemoryDatabase() {}

    /**
     * Write a field. Returns {@code ""}.
     */
    public String set(String key, String field, String value) {
        return "";
    }

    /**
     * Returns the value, or {@code ""} if missing.
     */
    public String get(String key, String field) {
        return "";
    }

    /**
     * Returns {@code "true"} if deleted, {@code "false"} if missing.
     */
    public String delete(String key, String field) {
        return "";
    }

    /**
     * Fields as {@code field(value), ...} sorted by field. Empty if none.
     * Example: {@code "abc(123), age(30), city(NY), name(Alice)"}.
     */
    public String scan(String key) {
        return "";
    }

    /**
     * Same as scan, only fields whose names start with prefix.
     * Example: {@code "abc(123), age(30)"}.
     */
    public String scanByPrefix(String key, String prefix) {
        return "";
    }

    /**
     * Write with no expiry. Returns {@code ""}.
     */
    public String setAt(String key, String field, String value, int timestamp) {
        return "";
    }

    /**
     * Write live in {@code [timestamp, timestamp + ttl)}. Returns {@code ""}.
     */
    public String setAtWithTtl(String key, String field, String value, int timestamp, int ttl) {
        return "";
    }

    /**
     * Delete if live at timestamp. Returns {@code "true"} or {@code "false"}.
     */
    public String deleteAt(String key, String field, int timestamp) {
        return "";
    }

    /**
     * Value if live at timestamp, else {@code ""}.
     */
    public String getAt(String key, String field, int timestamp) {
        return "";
    }

    /**
     * Live fields at timestamp as {@code field(value), ...}, or {@code ""}.
     */
    public String scanAt(String key, int timestamp) {
        return "";
    }

    /**
     * Live fields with that prefix at timestamp, or {@code ""}.
     */
    public String scanByPrefixAt(String key, String prefix, int timestamp) {
        return "";
    }

    /**
     * Snapshot live keys. Returns the key count as a string.
     */
    public String backup(int timestamp) {
        return "";
    }

    /**
     * Load the latest backup at or before timestampToRestore. Returns
     * {@code ""}.
     */
    public String restore(int timestamp, int timestampToRestore) {
        return "";
    }
}
